core/nodejsActionBase/runner.js

```diff
@@ -67,8 +67,7 @@ function NodeActionRunner() {
         } else {
             // The code is a plain old JS file.
             try {
-                eval(message.code);
-                thisRunner.userScriptMain = eval(message.main);
+                thisRunner.userScriptMain = eval('(function(){\n' + message.code + '\nreturn ' + message.main + '})()');
```

Comment:
I have also prototyped the file-based option, so I can easily switch to the alternative if desired.

However, require basically does the same thing as a function scope/eval except that it also adds additional parameters to the function, e.g., `module` and `exports`. These names did not exist until now. The file-based solution is fine and certainly more consistent with the zip flow, but at the same time a bigger departure from the current code. It is also bit trickier to get right as the action code could be already messing with `exports` and `module`. Specifically, I don't think `exports.main = $message.main` is defensive enough.

I don't know openwhisk enough yet to know if the action file name can reuse the action name (in particular if valid action name => valid file name). I suspect a temporary subdirectory is also needed to isolate the arbitrary named file and I guess the random folder will show in the stack trace.